### DISCUSSION TITLE
Trouble getting started with IB + Nautilus

### DISCUSSION THREAD

**Discussion by skalaydzhiyski** [nautechsystems/nautilus_trader#1245] [Q&A] [answered]
Hi all,

I really really like nautilus and what it enables me to do compared to my current setup with backtrader, but I cannot find a way to a working example for using IB's integration :(

I managed to get the Live demo from /examples working, but I cannot find anything else anywhere.

Basically I want to be able to do thse 3 simple things:
1. Look for instruments using the Instrument provider for IB (using the IB Gateway running locally , not within a container)
    and print them out or something.
2. Get historical houlry/daily data for an instrument (simple stock - GOOG for ex) for the last year or two
3. Back test the simplest possible strategy where I buy at the first period and sell at the last.

I don't know where to even start so any help is appreciated !

All the best

Example Code:
```python
from nautilus_trader.adapters.interactive_brokers.common import *
from nautilus_trader.adapters.interactive_brokers.config import *
from nautilus_trader.adapters.interactive_brokers.providers import InteractiveBrokersInstrumentProvider
from nautilus_trader.adapters.interactive_brokers.client.client import InteractiveBrokersClient

from nautilus_trader.msgbus.bus import MessageBus
from nautilus_trader.cache.cache import Cache
from nautilus_trader.common.logging import Logger
from nautilus_trader.common.clock import TestClock
from nautilus_trader.config import LoggingConfig
from nautilus_trader.config import RoutingConfig
from nautilus_trader.model.identifiers import TraderId

from nautilus_trader.config import InstrumentProviderConfig
from nautilus_trader.persistence.catalog import ParquetDataCatalog
from nautilus_trader.persistence.external.core import write_objects
from nautilus_trader.adapters.interactive_brokers.factories import (
    get_cached_ib_client, get_cached_interactive_brokers_instrument_provider
)

import asyncio


clock = TestClock()
logger = Logger(clock)
trader_id = TraderId('bt-1')
msgbus = MessageBus(trader_id, clock, logger)
cache = Cache(logger)

contract = IBContract(
    secIdType='CRYPTO',
    symbol='BTC',
    exchange='PAXOS',
    #primaryExchange='NASDAQ'
    currency='USD'
)
instrument_provider_config = InteractiveBrokersInstrumentProviderConfig(
    load_ids=frozenset(["BTC/USD.PAXOS"]),
    load_contracts=frozenset([contract]),
)

catalog = ParquetDataCatalog("./catalog/")

gateway_config = InteractiveBrokersGatewayConfig(
    start=False,
    username="user",
    password="pass",
    trading_mode="paper",
    read_only_api=False,
)

client = get_cached_ib_client(
    loop=asyncio.get_event_loop(),
    msgbus=msgbus,
    cache=cache,
    clock=clock,
    logger=logger,
    host='127.0.0.1',
    port=4002,
    client_id=2,
    gateway=gateway_config
)
instrument_provider = get_cached_interactive_brokers_instrument_provider(
    client=client,
    config=instrument_provider_config,
    logger=logger
)
res = instrument_provider.load_all()
```

Output
```
1970-01-01T00:00:00.000000000Z [INF] TRADER-000.Cache: READY.
2023-09-18T12:00:57.171911000Z [WRN] TRADER-000.InteractiveBrokersClient-002: Error 504 req_id=-1: Not connected
```

**Comment by filipmacek** [comment]
Good point, will update docs/README for this. Was just setting this up recently, so I am pretty familiar with interactive brokers setup, for me it works pretty well and nautilus IB adapter looks really good

**Reply by skalaydzhiyski** [reply]
Amazing, thank you for the fast response ! Pls let me know when the docs are updated I would love to play around with IB !

**Reply by skalaydzhiyski** [reply]
Hey @filipmacek , sorry to pester you again, but did you have a chance to look into updating the docs with examples on the above ? I'm still stuck in the same place :(

All the best,
Spas

**Reply by filipmacek** [reply]
Created a branch with some docs, try this quickstart 
https://github.com/nautechsystems/nautilus_trader/blob/interactive-brokers-docs/docs/integrations/ib.md#quickstart
After that you should run successfully `interactive_brokers_example.py` strategy what will print `EUR/USD` ticks
```
2023-09-21T15:30:59.614061000Z [INF] TESTER-001.SubscribeStrategy-000: EUR/USD.IDEALPRO,1.06595,1.06597,3500000.00,2000000.00,1695310259000000000
2023-09-21T15:30:59.614085000Z [INF] TESTER-001.SubscribeStrategy-000: EUR/USD.IDEALPRO,1.06595,1.06596,500000.00,2000000.00,1695310259000000000
```

**Reply by skalaydzhiyski** [reply]
Hey, thanks for coming back to me, I already had that working to be honest (albeit with a different example) and that's fine, but I was wondering how to get interrogate the different classes like for example:
* Instantiate a client -> then see what hte client can do (send an order, get account summary, etc..)
* Instantiate a InstrumentProvider -> then get a list of matching contracts etc..
* Instantiate DataClient -> get historical data or live quotes from IB

I want to be able to do this mainly in order to explore the Nautilus API in a notebook and learn how things work, not for doing backtests yet.

Maybe I am completely wrong and that's not how you're supposed to use the framework ?
Sorry if it's a silly question.

**Reply by filipmacek** [reply, marked as answer]
Best to explore all strategy examples, learn how to use `Strategy` class function and what are common function/patterns there. You can start playing with the nautilus adapters, how to request data, indicators, or print live WS updates. Crypto exchange adapters and their API documentation are , in my opinion, better and easier to use at the beginning to get a good feeling of how everything works. Start by implementing some simple strategies that you understand, and check frontend UI for live updates. For that, best are MM strategies or grids, to see in real-time how algorithm trades and executes orders.

**Reply by skalaydzhiyski** [reply]
Thank you, I will have a look !
